File: pages/refactorings/replace-mutable-fields-with-lenses/01-account-balance/Spec.scala
//> using scala 3.3.4
//> using dep qa.hedgehog::hedgehog-core:0.14.0
//> using dep qa.hedgehog::hedgehog-runner:0.14.0
import hedgehog.*, hedgehog.core.*, hedgehog.runner.*

object Props extends Properties:
  def tests: List[Test] = List(
    property("withdraw: Before == After", agrees),
    property("the balance lens obeys get-put, put-get, put-put", laws),
  )

  // Raw values, so the same input feeds both Account types.
  val genInt: Gen[Int] = Gen.int(Range.linear(-30, 30))

  def agrees: Property =
    for
      b  <- genInt.forAll
      bn <- genInt.forAll
      am <- genInt.forAll
    yield
      val before = Before.withdraw(Before.Account(b, bn), am)
      val after  = After.withdraw(After.Account(b, bn), am)
      (before.balance, before.bonus) ==== (after.balance, after.bonus)

  def laws: Property =
    for
      b  <- genInt.forAll
      bn <- genInt.forAll
      v1 <- genInt.forAll
      v2 <- genInt.forAll
    yield
      val a = After.Account(b, bn)
      val l = After.balance
      l.get(l.set(v1)(a)) ==== v1 and
        l.set(l.get(a))(a) ==== a and
        l.set(v2)(l.set(v1)(a)) ==== l.set(v2)(a)

@main def spec(): Unit =
  val results = Props.tests.map { t =>
    val r = Property.check(t.withConfig(PropertyConfig.default),
      t.result, Seed.fromTime())
    println(Test.renderReport(
      "Props", t, r, ansiCodesSupported = false))
    r.status
  }
  if !results.forall(_ == Status.ok) then sys.exit(1)

File: pages/refactorings/replace-mutable-fields-with-lenses/01-account-balance/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad (unless)
import System.Exit (exitFailure)
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified Before
import qualified After

genInt :: Gen Int
genInt = Gen.int (Range.linear (-30) 30)

prop_agrees :: Property
prop_agrees = property $ do
  b  <- forAll genInt
  bn <- forAll genInt
  am <- forAll genInt
  let before = Before.withdraw (Before.Account b bn) am
      after  = After.withdraw (After.Account b bn) am
  (Before.balance before, Before.bonus before)
    === (After.balance after, After.bonus after)

-- A lens has to satisfy the three equations; they are exactly what
-- makes replacing a field with a get/set pair a refactoring.
prop_laws :: Property
prop_laws = property $ do
  b  <- forAll genInt
  bn <- forAll genInt
  v1 <- forAll genInt
  v2 <- forAll genInt
  let l = After.balanceLens
      a = After.Account b bn
  After.set l (After.get l a) a === a
  After.get l (After.set l v1 a) === v1
  After.set l v2 (After.set l v1 a) === After.set l v2 a

main :: IO ()
main = do
  ok <- checkParallel $ Group "Props"
    [ ("withdraw: Before == After", prop_agrees)
    , ("balance lens obeys get-put, put-get, put-put", prop_laws)
    ]
  unless ok exitFailure
